### PR TITLE
ARROW-15854: [C++] Refine CSV writer code

### PR DESCRIPTION
**Pull request by cyb70289**
- populate columns in normal order
- support int64 row length

### CODE REVIEW

File: cpp/src/arrow/csv/writer.cc [hunk 1]
```diff
@@ -117,17 +116,16 @@ class ColumnPopulator {
...
+  // pointers in preparation for calls to other (next) ColumnPopulators.
   // Implementations may apply certain checks e.g. for illegal values, which in case of
   // failure causes this function to return an error Status.
   // Args:
   //   output: character buffer to write to.
-  //   offsets: an array of end of row column within the the output buffer (values are
-  //   one past the end of the position to write to).
-  virtual Status PopulateColumns(char* output, int32_t* offsets) const = 0;
+  //   offsets: an array of start of row column within the output buffer.
+  virtual Status PopulateRows(char* output, int64_t* offsets) const = 0;
```
Comment (cyb70289) [inline cpp/src/arrow/csv/writer.cc:125]: Name `PopulateColumns` is a bit confusing IMO as this function writes rows of a single column to csv output buffer.

**Comment by cyb70289** [comment]
Don't see obvious performance difference from this PR.

**Comment by cyb70289** [comment]
@ursabot please benchmark lang=C++

File: cpp/src/arrow/csv/writer.cc (hunk 1 above)
Comment (emkornfield) [reply cpp/src/arrow/csv/writer.cc:125]: The new name seems fine to me

File: cpp/src/arrow/csv/writer.cc [hunk 2]
```diff
@@ -136,15 +134,14 @@ class ColumnPopulator {
...
-// Returns the position prior to last copied character (out_end is decremented).
-char* EscapeReverse(arrow::util::string_view s, char* out_end) {
-  for (const char* val = s.data() + s.length() - 1; val >= s.data(); val--, out_end--) {
-    if (*val == '"') {
-      *out_end = *val;
-      out_end--;
+// Copies the contents of s to out properly escaping any necessary characters.
+// Returns the position next to last copied character.
+char* Escape(arrow::util::string_view s, char* out_end) {
+  for (const char c : s) {
```
Comment (emkornfield) [inline cpp/src/arrow/csv/writer.cc:140]: It occurs to me that performance here might be improved by using higher level methods to find the next escape character and memcpy in between.  Probably depends on string lengths being processed.

**Comment by emkornfield** [comment]
I'm slightly ambivalent on reversing the population direction, it seems border line like change just for the sake of change.  I don't recall if the reverse population was necessary for some reason that has changed or just an overcomplicating on my part.   If you think it makes things easier to understand I'll do a more a thorough review

File: cpp/src/arrow/csv/writer.cc (hunk 2 above)
Comment (cyb70289) [reply cpp/src/arrow/csv/writer.cc:140]: Yes. It's the main reason I change the scan order. This function is the hotspot of string with quotes benchmark. Reverse scanning makes it hard to improve, the output string header pos is not known until end.

**Comment by cyb70289** [comment]
> I'm slightly ambivalent on reversing the population direction, it seems border line like change just for the sake of change. I don't recall if the reverse population was necessary for some reason that has changed or just an overcomplicating on my part. If you think it makes things easier to understand I'll do a more a thorough review

Thanks @emkornfield. I agree change just for the sake of change is not good. As I commented above, this change may make further optimization easier. And I do think it simplifies the code. Reading input strings forward continuously is also more friendly to prefetcher, though this workload is cpu bound.

File: cpp/src/arrow/csv/writer.cc [hunk 3]
```diff
@@ -572,33 +558,34 @@ class CSVWriterImpl : public ipc::RecordBatchWriter {
...
         static_cast<int32_t>(batch.num_columns() - 1 + options_.eol.size());
-    offsets_[0] += delimiters_length;
-    for (int64_t row = 1; row < batch.num_rows(); row++) {
-      offsets_[row] += offsets_[row - 1] + delimiters_length;
+    int64_t last_row_length = offsets_[0] + delimiters_length;
+    offsets_[0] = 0;
+    for (size_t row = 1; row < offsets_.size(); ++row) {
+      const int64_t this_row_length = offsets_[row] + delimiters_length;
+      offsets_[row] = offsets_[row - 1] + last_row_length;
+      last_row_length = this_row_length;
```
Comment (cyb70289) [inline cpp/src/arrow/csv/writer.cc:568]: `offsets_[i]` was the length of row `i` before the conversion.
After the conversion, `offsets_[i]` is the start position of row `i` in output buffer.

File: cpp/src/arrow/csv/writer.cc [hunk 4]
```diff
@@ -309,40 +300,33 @@ class QuotedColumnPopulator : public ColumnPopulator {
...
+            row = Escape(s, row);
           }
-          *(row_end - next_column_offset) = '"';
-          *(row_end - end_chars_.size() - 1) = '"';
-          memcpy(row_end - end_chars_.size(), end_chars_.data(), end_chars_.length());
-          *offsets -= next_column_offset;
+          *row++ = '"';
+          memcpy(row, end_chars_.data(), end_chars_.length());
+          row += end_chars_.length();
+          *offsets += static_cast<int64_t>(row - (output + *offsets));
```
Comment (pitrou) [inline cpp/src/arrow/csv/writer.cc:320]: Wouldn't it be simpler as:
```suggestion
          *offsets = static_cast<int64_t>(row - output);
```

File: cpp/src/arrow/csv/writer.cc [hunk 5]
```diff
@@ -309,40 +300,33 @@ class QuotedColumnPopulator : public ColumnPopulator {
...
-          int32_t next_column_offset =
-              static_cast<int32_t>(null_string_->size() + end_chars_.size());
-          memcpy((output + *offsets - next_column_offset), null_string_->data(),
-                 null_string_->size());
-          memcpy((output + *offsets - end_chars_.size()), end_chars_.c_str(),
+          memcpy(output + *offsets, null_string_->data(), null_string_->size());
+          memcpy(output + *offsets + null_string_->size(), end_chars_.c_str(),
                  end_chars_.size());
-          *offsets -= next_column_offset;
+          *offsets += static_cast<int32_t>(null_string_->size() + end_chars_.size());
```
Comment (pitrou) [inline cpp/src/arrow/csv/writer.cc:329]: `int64_t` perhaps?

File: cpp/src/arrow/csv/writer.cc (hunk 3 above)
Comment (pitrou) [reply cpp/src/arrow/csv/writer.cc:568]: Can you add a comment with this?

Comment (pitrou) [reply cpp/src/arrow/csv/writer.cc:568]: Also, you probably mean: "`offsets_[i]` was the length of row `i-1` before the conversion"...?

**Review by pitrou — COMMENTED** [review]
I think this is a welcome change. Just a couple comments.

File: cpp/src/arrow/csv/writer.cc [hunk 6]
```diff
@@ -136,15 +134,14 @@ class ColumnPopulator {
...
-// Copies the contents of to out properly escaping any necessary characters.
-// Returns the position prior to last copied character (out_end is decremented).
-char* EscapeReverse(arrow::util::string_view s, char* out_end) {
-  for (const char* val = s.data() + s.length() - 1; val >= s.data(); val--, out_end--) {
-    if (*val == '"') {
-      *out_end = *val;
-      out_end--;
+// Copies the contents of s to out properly escaping any necessary characters.
+// Returns the position next to last copied character.
+char* Escape(arrow::util::string_view s, char* out_end) {
```
[the remark this answers is not in the document]
Comment (cyb70289) [reply cpp/src/arrow/csv/writer.cc:139]: changed

File: cpp/src/arrow/csv/writer.cc (hunk 4 above)
Comment (cyb70289) [reply cpp/src/arrow/csv/writer.cc:320]: Yeah

File: cpp/src/arrow/csv/writer.cc [hunk 7]
```diff
@@ -184,28 +181,23 @@ class UnquotedColumnPopulator : public ColumnPopulator {
...
-      int32_t next_column_offset =
-          static_cast<int32_t>(null_string_->size() + end_chars_.size());
-      memcpy((output + *offsets - next_column_offset), null_string_->data(),
-             null_string_->size());
-      memcpy((output + *offsets - end_chars_.size()), end_chars_.c_str(),
+      memcpy(output + *offsets, null_string_->data(), null_string_->size());
+      memcpy(output + *offsets + null_string_->size(), end_chars_.c_str(),
              end_chars_.size());
-      *offsets -= next_column_offset;
+      *offsets += static_cast<int32_t>(null_string_->size() + end_chars_.size());
```
[the remark this answers is not in the document]
Comment (cyb70289) [reply cpp/src/arrow/csv/writer.cc:200]: changed

File: cpp/src/arrow/csv/writer.cc (hunk 5 above)
Comment (cyb70289) [reply cpp/src/arrow/csv/writer.cc:329]: changed

File: cpp/src/arrow/csv/writer.cc (hunk 3 above)
Comment (cyb70289) [reply cpp/src/arrow/csv/writer.cc:571]: comment added

**Review by emkornfield — APPROVED** [review]

File: cpp/src/arrow/csv/writer.cc [hunk 8]
```diff
@@ -571,34 +556,39 @@ class CSVWriterImpl : public ipc::RecordBatchWriter {
           column_populators_[col]->UpdateRowLengths(*batch.column(col), offsets_.data()));
     }
     // Calculate cumulative offsets for each row (including delimiters).
-    // ',' * num_columns - 1(last column doesn't have ,) + eol
-    int32_t delimiters_length =
+    // - before conversion: offsets_[i] = length of i-th row
```
Comment (pitrou) [inline cpp/src/arrow/csv/writer.cc:559]:
```suggestion
    // - before conversion: offsets_[i] = length of (i-1)th row
```

File: cpp/src/arrow/csv/writer.cc [hunk 9]
```diff
@@ -571,34 +556,39 @@ class CSVWriterImpl : public ipc::RecordBatchWriter {
...
+    // - after conversion:  offsets_[i] = offset to the starting of i-th row buffer
+    //   - offsets_[0] = 0
+    //   - offsets_[i] = offsets_[i-1] + len(i-1-th row) + len(delimiters)
+    // Delimiters: ',' * num_columns - 1(last column doesn't have ,) + eol
```
Comment (pitrou) [inline cpp/src/arrow/csv/writer.cc:563]:
```suggestion
    // Delimiters: ',' * (num_columns - 1) + eol
```

**Review by pitrou — COMMENTED** [review]
Just two nits, LGTM otherwise.

File: cpp/src/arrow/csv/writer.cc (hunk 8 above)
Comment (cyb70289) [reply cpp/src/arrow/csv/writer.cc:559]: `offsets_[i]` is indeed the length of  `i-th` row before conversion, not `(i-1)th`.

Comment (pitrou) [reply cpp/src/arrow/csv/writer.cc:559]: Oops, my bad.

**Review by pitrou — APPROVED** [review]

**Comment by cyb70289** [comment]
CI errors not related

**Pull request closed without merging**